### PR TITLE
use builders to create connections with sharding keys

### PR DESCRIPTION
**Pull request by njr-11**
Add code to utilize the ConnectionBuilder/PooledConnectionBuilder/XAConnectionBuilder, depending on which type of data source, when a connection is requested with a sharding key and a matching connection does not already exist in the pool.
Add basic test cases of connection sharing, and at least one that verifies that an existing connection is matched from the free pool.

### CODE REVIEW

File: dev/com.ibm.ws.jdbc_fat_v43/test-applications/app43/src/jdbc/fat/v43/web/JDBC43TestServlet.java [hunk 1]
```diff
@@ -1669,6 +1789,93 @@ public void testSuspendRunLTCAndResumeUnsharable() throws Exception {
...
+            con4.close();
+        }
+
+        // Specify a match for the first sharding key
+        ShardingKey key5 = keybuilderB.build();
+        Connection con5 = conbuilderB.shardingKey(key5).build();
+        try {
+            assertEquals(Connection.TRANSACTION_READ_COMMITTED, con5.getTransactionIsolation());
+            String k5 = con5.getClientInfo("SHARDING_KEY"); // extension by mock JDBC driver to determine the sharding key
+            assertTrue(k5, k5.endsWith("|CHAR:XACBKey;BOOLEAN:true;"));
```
Comment (alexsm82) [inline dev/com.ibm.ws.jdbc_fat_v43/test-applications/app43/src/jdbc/fat/v43/web/JDBC43TestServlet.java]: Weren't there no subkeys specified on this connection?

Comment (njr-11) [reply dev/com.ibm.ws.jdbc_fat_v43/test-applications/app43/src/jdbc/fat/v43/web/JDBC43TestServlet.java]: The subkeys were set upon previous usage of the builder:
ShardingKey keyB = keybuilderB.subkey("XACBKey", JDBCType.CHAR).subkey(true, JDBCType.BOOLEAN).build();
Note that the JDBC 4.3 spec is vague about what happens when a ShardingKeyBuilder is reused -- whether it rejects reuse, allows reuse with subkeys preserved, allows reuse with subkeys cleared, or something else.  Our mock JDBC driver happens to allow reuse with the subkeys preserved, which is the behavior that this test case relies on here.

**Comment by njr-11** [comment]
#build

**Comment by njr-11** [comment]
#build

**Comment by njr-11** [comment]
Build has 3 test failures which are duplicates of pre-existing issues in unrelated test buckets.
https://wasrtc.hursley.ibm.com:9443/jazz/resource/itemOid/com.ibm.team.build.BuildResult/_zh_BwLzPEeieisOzGH_nZQ

**Pull request merged**